### ISSUE TITLE
method override acts as escape hatch for `@experimental`

### ISSUE THREAD

**Issue by bishabosha** [lampepfl/dotty#17291]
## Compiler version

3.3.0-RC3

## Minimized code

```Scala
import annotation.experimental

trait Foo { @experimental def bar = 23 }

object Bar extends Foo { override def bar = 42 }

@main def Test = println(Bar.bar)
```

## Output

```scala
42
```

## Expectation

I expected an error at the use site of `Bar.bar`.

Perhaps also we should error at the definition when overriding experimental definitions to also add experimental.

**Comment by bishabosha** [comment]
this one is not a bug, due to rules of JVM

**Comment by SethTisue** [comment]
> Perhaps also we should error at the definition when overriding experimental definitions to also add experimental

seems like a good idea to me

**Comment by dwijnand** [comment]
> > Perhaps also we should error at the definition when overriding experimental definitions to also add experimental
> 
> seems like a good idea to me

I disagree.  Imagine Bar is `scala.tools.dotc.internal.CompilerImpl` and Foo is `scala.tools.dotc.interface.Compiler` and exposes its method as `@experimental`, then I shouldn't have to pollute `CompilerImpl`.

**Comment by SethTisue** [comment]
Hmm.... I had not understood the meaning of `@experimental` to include a use case such as that. Experimental != internal. Am I out of step on this...?

**Comment by SethTisue** [comment]
Hmmm.... well that gives me pause, but doesn't necessarily convince me. Presumably `CompilerImpl` would be package-private. Perhaps the rule @bishabosha proposes only makes sense for public methods?

I admit to being unfamiliar with how `@experimental` already interacts with visibility, if it does interact.

**Comment by bishabosha** [comment]
So extending a trait with an experimental method should now commit to supporting that new method forever, even if the trait eventually removes the experimental method?

**Comment by sjrd** [comment]
> So a class that extends a trait (where the trait has an experimental method) should now commit to supporting that new method forever, even if the trait eventually removes the experimental method? (because it replaced by a different design)

Yes, absolutely. If it does not want to commit to it, the subclass should also mark it `@experimental`.

**Comment by bishabosha** [comment]
> > So a class that extends a trait (where the trait has an experimental method) should now commit to supporting that new method forever, even if the trait eventually removes the experimental method? (because it replaced by a different design)
> 
> Yes, absolutely. If it does not want to commit to it, the subclass should also mark it `@experimental`.

Ok so e.g. if we add `@experimental def defaultArgument(index: Int): Any` to `deriving.Mirror.Product` then the synthetic override in companion objects should probably declare it experimental also

**Comment by SethTisue** [comment]
I see y'all's point of view, but it just seems to me like it's expecting a lot of library authors to remember to do this. Sure, if we bites them, then afterwards we could say "we told, right there in plain sight on issue 17291, behind the sign saying 'beware of the leopard', didn't you read it? everyone knows issue 17291!" I'm having trouble feeling satisfied with this.

What about a warning, that library authors could choose to silence with `@nowarn`?

**Comment by nicolasstucki** [comment]
With the warning, we would need to write
```scala
package foo
trait Foo:
  @deprecated("Use Bar.f instead") def f: Int
trait Bar extends Foo:
  @nowarn("""cat=deprecation&origin=foo\.Foo""")
  def f = 42
```
